RFC 0002: Textures Description

tnqzh123 opened the pull request:
[\>\> 点击查看全文 \<\<](https://github.com/bs-community/rfcs/blob/master/rfcs/0002-textures-desc.md)

tnqzh123 committed:
add textures-desc

tnqzh123 committed:
update

tnqzh123 committed:
update

g-plane commented:
1. 即使 TID 具有唯一性，最好还是单独设置一个 `id` 字段作为真正的主键，而 `tid` 作为外键，这样也符合一般的数据库设计；
2. options 表实际上除了 `id` 作为主键是 `int` 之外，其它都是 `varchar`；
3. 可以考虑在材质描述的右上角添加一个「编辑」按钮，更直观也更方便；
4. 关于转义的问题，这个可以考虑使用 [thephpleague/commonmark](https://github.com/thephpleague/commonmark)，在里面设置好 GitHub Flavored Markdown 可以使得转义危险标签的同时，正常显示普通的标签。

tnqzh123 committed:
update

tnqzh123 commented:
@g-plane 

> 1. 即使 TID 具有唯一性，最好还是单独设置一个 `id` 字段作为真正的主键，而 `tid` 作为外键，这样也符合一般的数据库设计；
> 2. options 表实际上除了 `id` 作为主键是 `int` 之外，其它都是 `varchar`；
> 3. 可以考虑在材质描述的右上角添加一个「编辑」按钮，更直观也更方便；
> 4. 关于转义的问题，这个可以考虑使用 [thephpleague/commonmark](https://github.com/thephpleague/commonmark)，在里面设置好 GitHub Flavored Markdown 可以使得转义危险标签的同时，正常显示普通的标签。

关于第三点，其实一开始我也是这么想的，但是考虑到有的用户在发布材质时可能不会填写材质描述，而等到材质发布后再增加描述的情况，因为不存在材质描述时不应该展示材质描述的 `card`，也就不会显示这个「编辑」按钮。所以还是考虑增加一个 `btn` 来修改材质描述，对用户来说也更显眼。

g-plane commented:
那……要不要考虑两个都加？

tnqzh123 commented:
@g-plane 

> 那……要不要考虑两个都加？

我觉得没有必要...？

「编辑描述」的 `btn` 我觉得够显眼了。

![image](https://user-images.githubusercontent.com/16630630/87758299-cbbe5f00-c83e-11ea-808f-90a531790653.png)

g-plane commented:
还可以。

mochaaP commented:
@tnqzh123 https://codepen.io/mochaa/pen/BajqMvQ 我的思路是这样。

mochaaP commented:
@mochaaP 🤔 刚才 CodePen 好像没同步上，现在应该可以了

tnqzh123 commented:
@mochaaP 

> https://codepen.io/mochaa/pen/BajqMvQ 我的思路是这样。

当材质存在描述，或当前用户为材质上传者时，展示材质描述的 `card` 吗？

好像也不错。

g-plane commented:
1. Markdown 那里应该是开启 GFM 扩展，因为 html_input 会把所有 HTML 标签都处理掉；
2. 我们打算采纳 @mochaaP 的建议吗？如果是，也应该更新 RFC。
3. 确定是采用插件的方案吗？

tnqzh123 committed:
update

tnqzh123 commented:
@g-plane 

> 1. Markdown 那里应该是开启 GFM 扩展，因为 html_input 会把所有 HTML 标签都处理掉；
> 2. 我们打算采纳 @mochaaP 的建议吗？如果是，也应该更新 RFC。
> 3. 确定是采用插件的方案吗？

1. Updated；
2. Updated；
3. 我比较倾向于插件，这样可以更灵活，技术上应该也没有什么难点吧？

mochaaP commented on rfcs/0002-textures-desc.md:
```diff
@@ -32,17 +32,17 @@
 
 ### 材质描述的字数限制
 
 材质描述应存在字数限制，但这个限制应该允许管理员在插件配置中自行设定，且应该在用户填写材质描述时提示（在文本框下方增加一个 `callout-info` 或 `alert-info`）。前后端都应该对字数限制进行检查，若字数超限，则弹出 modal 或 toast 来提示用户。
```
我认为使用 maxlength 属性规定（+ <https://github.com/mimo84/bootstrap-maxlength> 这个库的样式或许可以再改改）更好。

tnqzh123 replied:
@mochaaP 我个人认为 maxlength 属性有在一定程度上影响用户体验（写满之后就不能继续写了，会打断用户的写作），不过 bootstrap-maxlength 如果可以在 textarea 没有指定 maxlength 属性的情况下也可以显示的话，那是可以考虑的

mochaaP commented on rfcs/0002-textures-desc.md:
```diff
@@ -32,17 +32,17 @@
+可以在材质详情页中的材质预览的下方、评论区（如果有）的上方增加一个 `card-secondary` 来展示。当材质描述不存在时，除非当前用户为该材质的上传者或管理员，否则不应该展示材质描述。
 
-应该支持解析 Markdown。可以使用 [thephpleague/commonmark](https://github.com/thephpleague/commonmark) 来解析 GitHub Flavored Markdown，且将 `html_input` 参数和 `allow_unsafe_links` 参数的值分别设为 `strip` 和 `false`，以防范 XSS 攻击。
+应该支持解析 Markdown。可以使用 [thephpleague/commonmark](https://github.com/thephpleague/commonmark) 开启 GFM 扩展来解析 GitHub Flavored Markdown。
```
是否给用户回传 Markdown 以便 in-place 编辑？

tnqzh123 replied:
@mochaaP

> 是否给用户回传 Markdown 以便 in-place 编辑？

这不是理所当然的嘛…

mochaaP reviewed (changes requested):
🤔 还需要规定存储和 API 的细节。

tnqzh123 committed:
add API description

tnqzh123 commented:
@mochaaP 

> 🤔 还需要规定存储和 API 的细节。

存储需要规定什么细节？我觉得直接存进来就可以了吧，当然要做好反 SQL 注入什么的。

API 细节补了。

g-plane commented on rfcs/0002-textures-desc.md:
```diff
@@ -28,6 +28,14 @@
+
+所有 API 都要求用户必须登录后才可请求。
+
+获取材质描述：`GET /skinlib/desc/{tid}`，无请求体；
```
建议用这样的 URL：`/textures/{tid}/descrption`。

g-plane commented on rfcs/0002-textures-desc.md:
```diff
@@ -0,0 +1,79 @@
+
+获取材质描述：`GET /skinlib/desc/{tid}`，无请求体；
+
+更新材质描述：`POST /skinlib/desc/{tid}`，请求体内容为材质描述文本。
```
建议使用 PUT 而不是 POST。

mochaaP reviewed (approved):
LGTM.

tnqzh123 committed:
update APIs

tnqzh123 committed:
update API url

g-plane commented:
前面显示「添加描述」的按钮跟后面「如果没有描述就不显示」是不是有冲突？
另外，如果采用 Codepen 例子那样的方案，要不要将按钮文本统一显示为「编辑描述」？（不管有没有描述）

tnqzh123 commented:
@g-plane 

> 前面显示「添加描述」的按钮跟后面「如果没有描述就不显示」是不是有冲突？
> 另外，如果采用 Codepen 例子那样的方案，要不要将按钮文本统一显示为「编辑描述」？（不管有没有描述）

不冲突的，因为「当前用户不是该材质的上传者或管理员」才「如果没有描述就不显示」。

Codepen 示例的方案中，当用户添加材质描述后，「添加描述」的按钮会消失，右上角会出现一个 `fa-edit`（但没有文字），所以我觉得应该是不存在统一按钮文本的问题的。Codepen 的预览是可以操作的，您可以自己试一试。

g-plane committed:
add implementation PR
